### ISSUE TITLE
Fatal error: exception File "typing/typeclass.ml", line 294

### ISSUE THREAD

**Issue by vicuna** [ocaml/ocaml#7894]
**Original bug ID:** 7894
**Reporter:** wlitwin
**Assigned to:** @lpw25
**Status:** assigned (set by @lpw25 on 2019-02-15T10:13:47Z)
**Resolution:** open
**Priority:** normal
**Severity:** crash
**Version:** 4.07.0
**Category:** typing
**Monitored by:** @nojb @gasche

## Bug description

OCaml 4.07.0 fails to compile the following program:

```ocaml
class basicWidget app = object(self)
    method invalidate : unit =
        app#redrawWidget self
end

class simple app =
    let label1 = new basicWidget app in
object(self)
    inherit basicWidget app
end

let _ = new simple (object end)
```

## Steps to reproduce

paste the description code into a src file and try building with:

```
ocamlc file.ml
```

The following happens:

```
Fatal error: exception File "typing/typeclass.ml", line 294, characters 12-18: Assertion failed
```

## File attachments

- [main.ml](https://gist.githubusercontent.com/vicuna/bb5fb9d3a41986ee9201730f55ddb39f/raw/3de80c24a0b196138404be8242d7bef9e97db170/main.ml)

**Comment by vicuna** [comment]
**Comment author:** wlitwin

Forgot to check trunk, still appears to throw an exception as of commit:

711362988734ea03b81c4ac9a32e07133572bc08
Date:   Tue Jan 8 17:20:42 2019 +0100

$ ./ocamlc --version
4.08.0+dev0-2018-04-09

$ ./ocamlc main.ml
Fatal error: exception File "typing/typeclass.ml", line 302, characters 15-21: Assertion failed

**Comment by vicuna** [comment]
**Comment author:** @gasche

This appears to be a regression in 4.07.0. Indeed, 4.06.1 prints the following message instead:

File "test.ml", line 10, characters 24-27:
Error: This expression has type < redrawWidget : basicWidget -> unit; .. >
       but an expression was expected of type
         < redrawWidget : (< invalidate : unit; .. > as 'a) -> unit; .. >
       Type basicWidget = < invalidate : unit > is not compatible with type
         < invalidate : unit; .. > as 'a 
       Self type cannot be unified with a closed object type

**Comment by vicuna** [comment]
**Comment author:** @gasche

I traced the regression back to this commit

  "make the dummy method disappear when finalizing a class"
  86e1c8a4e0cd8f52cd3ebdf57e8776ca79235c02
  https://github.com/ocaml/ocaml/commit/86e1c8a4e0cd8f52cd3ebdf57e8776ca79235c02

**Comment by vicuna** [comment]
**Comment author:** @trefis

We looked at this yesterday with Leo, and found a couple of other related issues.
Leo is currently working on a fix. (details to follow in a GPR)

**Comment by vicuna** [comment]
**Comment author:** @garrigue

Is there any progress on this one?
Interactions in the object system are subtle, as this problem has demonstrated.

**Comment by vicuna** [comment]
**Comment author:** @lpw25

I've got a patch coming that tidies up a fair few things in typeclass.ml and fixes this bug. Should take me a few more days, so assuming no urgent interruptions I should make a PR this week or next.

**Comment by trefis** [comment]
IIRC « (details to follow in a GPR) » was a forward reference to #8516 .

**Comment by trefis** [comment]
There was some progress on #8516 recently.

**Comment by trefis** [comment]
This was fixed by #8516: on 4.14 I get the same error message as on 4.06.
